File: crm-app/src/main/java/com/hotelcrm/crmapp/service/InteractionService.java
package com.hotelcrm.crmapp.service;

import com.hotelcrm.crmapp.dto.interaction.request.InteractionCompleteRequest;
import com.hotelcrm.crmapp.dto.interaction.request.InteractionCreateRequest;
import com.hotelcrm.crmapp.dto.interaction.request.InteractionFilterRequest;
import com.hotelcrm.crmapp.dto.interaction.request.InteractionUpdateRequest;
import com.hotelcrm.crmapp.dto.interaction.response.InteractionResponse;
import com.hotelcrm.crmapp.entity.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface InteractionService {

    InteractionResponse schedule(InteractionCreateRequest req, User currentUser);

    InteractionResponse complete(Long id, InteractionCompleteRequest req, User currentUser);

    InteractionResponse update(Long id, InteractionUpdateRequest req, User currentUser);

    Page<InteractionResponse> getFiltered(InteractionFilterRequest filter, Pageable pageable, Long userId);

    Page<InteractionResponse> calendar(Long userId, LocalDateTime from, LocalDateTime to, Pageable pageable);

    List<InteractionResponse> upcomingFollowUps(Long userId);

    Optional<InteractionResponse> getById(Long id, User currentUser);

    Long delete(Long id, User currentUser);
}

File: crm-frontend/src/app/shared/services/interaction.service.ts
import { Injectable } from '@angular/core';
import {environment} from '../../../environments/environment';
import {HttpClient, HttpParams} from '@angular/common/http';
import {
  InteractionCompleteRequest,
  InteractionCreateRequest,
  InteractionFilter,
  InteractionUpdateRequest
} from '../models/request/interaction-requests';
import {Observable} from 'rxjs';
import {Page} from '../models/helpers';
import {Interaction} from '../models/interaction.model';

@Injectable({ providedIn: 'root' })
export class InteractionService {
  private readonly baseUrl = `${environment.apiUrl}/interactions`;

  constructor(private readonly http: HttpClient) {}

  schedule(req: InteractionCreateRequest): Observable<Interaction> {
    return this.http.post<Interaction>(this.baseUrl, req, { withCredentials: true });
  }

  filter(filter: InteractionFilter, page = 0, size = 10): Observable<Page<Interaction>> {
    let params = new HttpParams().set('page', page).set('size', size);
    Object.entries(filter || {}).forEach(([k, v]) => {
      if (v !== undefined && v !== null && `${v}`.trim() !== '') params = params.set(k, `${v}`);
    });
    return this.http.get<Page<Interaction>>(`${this.baseUrl}/filter`, { params, withCredentials: true });
  }

  getById(id: number): Observable<Interaction> {
    return this.http.get<Interaction>(`${this.baseUrl}/${id}`, { withCredentials: true });
  }

  update(id: number, req: InteractionUpdateRequest): Observable<Interaction> {
    return this.http.put<Interaction>(`${this.baseUrl}/${id}`, req, { withCredentials: true });
  }

  complete(id: number, req: InteractionCompleteRequest): Observable<Interaction> {
    return this.http.patch<Interaction>(`${this.baseUrl}/${id}/complete`, req, { withCredentials: true });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`, { withCredentials: true });
  }

  calendar(from: string, to: string, page = 0, size = 10): Observable<Page<Interaction>> {
    let params = new HttpParams().set('from', from).set('to', to).set('page', page).set('size', size);
    return this.http.get<Page<Interaction>>(`${this.baseUrl}/calendar`, { params, withCredentials: true });
  }

  upcomingFollowUps(): Observable<Interaction[]> {
    return this.http.get<Interaction[]>(`${this.baseUrl}/follow-ups`, { withCredentials: true });
  }
}
